airflow/executors/kubernetes_executor.py

```diff
@@ -848,10 +850,13 @@ def sync(self) -> None:
 
     def _change_state(self, key: TaskInstanceKeyType, state: Optional[str], pod_id: str) -> None:
         if state != State.RUNNING:
-            if self.kube_config.delete_worker_pods:
+            if self.kube_config.delete_on_success and state is State.SUCCESS:
                 if not self.kube_scheduler:
                     raise AirflowException(NOT_STARTED_MESSAGE)
                 self.kube_scheduler.delete_pod(pod_id)
+            elif self.kube_config.delete_worker_pods:
+                if not self.kube_scheduler:
+                    raise AirflowException(NOT_STARTED_MESSAGE)
```

Comment:
Hm, is `self.kube_scheduler.delete_pod(pod_id)` missing in the `elif` branch?

Reply:
Yup it is